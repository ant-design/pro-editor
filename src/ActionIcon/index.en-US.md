---
title: ActionIcon
group: Basic
atomId: ActionIcon
demo:
  cols: 2
---

# ActionIcon

## When to use

It is a encapsulation of antd Button, allowing icons to have responsive button actions.

## Code Example

<code src="./demos/basic.tsx" ></code> <code title='Default' src="./demos/preset.tsx" description='Presets for some commonly used buttons in the editor'></code>

## API

| Parameter | Description    | Type                                       | Default |
| :-------- | :------------- | :----------------------------------------- | :------ |
| cursor    | Mouse type     | `CSSProperties['cursor']`                  | -       |
| title     | Action hint    | `TooltipProps['title']`                    | -       |
| placement | Hint position  | `TooltipProps['placement']`                | -       |
| icon      | Icon           | `ButtonProps['icon']`                      | -       |
| onClick   | Click callback | `ButtonProps['onClick']`                   | -       |
| size      | Icon size      | `'default' \| 'large' \|'small' \| number` | small   |

Refer to other API in antd Button Props.
