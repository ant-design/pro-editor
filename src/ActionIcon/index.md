---
title: ActionIcon 动作图标
group: 基础组件
atomId: ActionIcon
nav:
  title: 组件
  order: 2
---

# ActionIcon 动作图标

## 何时使用

对 antd Button 的封装，让图标具备按钮动作的响应式效果。

## 代码演示

### Basic

<code src="./demos/basic.tsx" ></code>

### 预设

一些编辑器常用按钮的预设

<code src="./demos/preset.tsx" ></code>

## API

| 参数      | 说明     | 类型                                       | 默认值 |
| :-------- | :------- | :----------------------------------------- | :----- |
| cursor    | 鼠标类型 | `CSSProperties['cursor']`                  | -      |
| title     | 动作提示 | `TooltipProps['title']`                    | -      |
| placement | 提示位置 | `TooltipProps['placement']`                | -      |
| icon      | 图标     | `ButtonProps['icon']`                      | -      |
| onClick   | 点击回调 | `ButtonProps['onClick']`                   | -      |
| size      | 图标尺寸 | `'default' \| 'large' \|'small' \| number` | small  |

其他 API 参考 antd Button Props .
