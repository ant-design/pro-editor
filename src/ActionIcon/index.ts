import ActionIcon from './ActionIcon';

export { default as ActionIcon } from './ActionIcon';
export type { ActionIconProps } from './ActionIcon';
export * from './Icons';
// 内部使用统一图标语义
export type { IconsProps } from './Icons';
export default ActionIcon;
