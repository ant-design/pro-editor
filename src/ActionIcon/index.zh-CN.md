---
title: ActionIcon 动作图标
group: 基础组件
atomId: ActionIcon
nav:
  title: 基础组件
  order: 2
demo:
  cols: 2
---

# ActionIcon 动作图标

## 何时使用

对 antd Button 的封装，让图标具备按钮动作的响应式效果。

## 代码演示

<code src="./demos/basic.tsx" ></code>
<code src="./demos/preset.tsx" ></code>

## API

| 参数      | 说明     | 类型                                       | 默认值 |
| :-------- | :------- | :----------------------------------------- | :----- |
| cursor    | 鼠标类型 | `CSSProperties['cursor']`                  | -      |
| title     | 动作提示 | `TooltipProps['title']`                    | -      |
| placement | 提示位置 | `TooltipProps['placement']`                | -      |
| icon      | 图标     | `ButtonProps['icon']`                      | -      |
| onClick   | 点击回调 | `ButtonProps['onClick']`                   | -      |
| size      | 图标尺寸 | `'default' \| 'large' \|'small' \| number` | small  |

其他 API 参考 antd Button Props .
