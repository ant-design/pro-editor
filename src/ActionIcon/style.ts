import { createStyles } from '../theme';

export const useStyles = createStyles(({ token, css, cx, prefixCls }, { size, className }) => {
  const prefix = `${prefixCls}-${token.editorPrefix}-icon`;
  const sizeBoundary =
    typeof size === 'number'
      ? css`
          width: ${size}px !important;
          height: ${size}px !important;
        `
      : '';

  const button = css`
    display: flex;
    align-items: center;
    justify-content: center;

    &:hover {
      color: ${token.colorText} !important;
    }

    &:active {
      scale: 0.8;
      color: ${token.colorText};
    }

    transition: color 600ms ${token.motionEaseOut}, scale 400ms ${token.motionEaseOut},
      background-color 100ms ${token.motionEaseOut};
  `;

  return {
    container: cx(prefix, button, sizeBoundary, className),
    tooltip: css`
      pointer-events: none;
    `,
  };
});
