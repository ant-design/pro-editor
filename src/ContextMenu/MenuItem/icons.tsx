const strokeWidth = 2.2;
const iconSize = '13px';

export const CommandIcon = () => (
  <svg
    fill="none"
    height={iconSize}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={strokeWidth}
    viewBox="0 0 24 24"
    width={iconSize}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M18 3a3 3 0 0 0-3 3v12a3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3H6a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3V6a3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3h12a3 3 0 0 0 3-3 3 3 0 0 0-3-3z"></path>
  </svg>
);
export const ControlIcon = () => (
  <svg
    className="lucide lucide-chevron-up"
    fill="none"
    height={iconSize}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={strokeWidth}
    viewBox="0 0 24 24"
    width={iconSize}
    xmlns="http://www.w3.org/2000/svg"
  >
    <polyline points="18 15 12 9 6 15"></polyline>
  </svg>
);
export const ShiftIcon = () => (
  <svg
    className="lucide lucide-chevron-up"
    fill="none"
    height={iconSize}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={strokeWidth}
    viewBox="0 0 24 24"
    width={iconSize}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M9 18v-6H5l7-7 7 7h-4v6H9z"></path>
  </svg>
);

export const AltIcon = () => (
  <svg
    className="lucide lucide-chevron-up"
    fill="none"
    height={iconSize}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={strokeWidth}
    viewBox="0 0 24 24"
    width={iconSize}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M3 3h6l6 18h6"></path>
    <path d="M14 3h7"></path>
  </svg>
);
