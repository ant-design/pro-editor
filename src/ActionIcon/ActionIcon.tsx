import type { ButtonProps, TooltipProps } from 'antd';
import { Button, Tooltip } from 'antd';
import type { CSSProperties, FC } from 'react';
import { ConfigProvider } from '../ConfigProvider';
import { useStyles } from './style';

/**
 * @title 动作图标属性
 * @description 继承自 `Button` 组件所有属性，除了 `title` 和 `size`
 */
export interface ActionIconProps extends Omit<ButtonProps, 'title' | 'size'> {
  /**
   * @title 鼠标类型
   */
  cursor?: CSSProperties['cursor'];
  /**
   * @title 动作提示
   */
  title?: TooltipProps['title'];
  /**
   * @title 提示位置
   */
  placement?: TooltipProps['placement'];
  /**
   * @title 图标
   */
  icon: ButtonProps['icon'];
  /**
   * @title 点击回调
   */
  onClick?: ButtonProps['onClick'];
  /**
   * @title 图标尺寸
   */
  size?: 'default' | 'large' | number;
  /**
   * @description 鼠标移入时候的延迟tooltip时间，默认 0.5
   * @default 0.5
   */
  tooltipDelay?: number;
  /**
   * @description 是否展示小箭头，默认不展示
   * @default false
   */
  arrow?: boolean;
}

const BaseActionIcon: FC<ActionIconProps> = ({
  placement,
  title,
  icon,
  cursor,
  onClick,
  className,
  arrow = false,
  size = 'default',
  tooltipDelay = 0.5,
  ...restProps
}) => {
  const { styles, cx } = useStyles({ size });

  const Icon = (
    <Button
      icon={icon}
      className={cx(styles.container, className)}
      type={'text'}
      style={{ cursor }}
      size={typeof size === 'number' || size === 'default' ? 'middle' : size}
      {...restProps}
      onClick={onClick}
    />
  );

  return (
    <>
      {!title ? (
        Icon
      ) : (
        <Tooltip
          arrow={arrow}
          overlayClassName={styles.tooltip}
          title={title}
          mouseEnterDelay={tooltipDelay}
          placement={placement}
        >
          {Icon}
        </Tooltip>
      )}
    </>
  );
};

const ActionIcon = (props: ActionIconProps) => {
  const { size } = props || {};
  const { theme: token } = useStyles({ size });

  return (
    <ConfigProvider
      componentToken={{
        Button: {
          colorText: token.colorTextTertiary,
          // TODO： Token 的提供不太合理，需要改造
          colorBgTextHover: token.colorFillSecondary,
          colorBgTextActive: token.colorFill,
        },
      }}
    >
      <BaseActionIcon {...props} />
    </ConfigProvider>
  );
};
export default ActionIcon;
