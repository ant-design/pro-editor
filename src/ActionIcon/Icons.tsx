import { DeleteFilled, EditFilled } from '@ant-design/icons';
import type { FC } from 'react';

import type { ActionIconProps } from './ActionIcon';
import ActionIcon from './ActionIcon';

export type IconsProps = Omit<ActionIconProps, 'icon'>;

const HandleIcon = (
  <svg viewBox="0 0 20 20" width="16" fill="currentColor">
    <path d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z" />
  </svg>
);

const CollapseIcon = (
  <svg width="10" viewBox="0 0 70 41" fill="currentColor">
    <path d="M30.76 39.2402C31.885 40.3638 33.41 40.995 35 40.995C36.59 40.995 38.115 40.3638 39.24 39.2402L68.24 10.2402C69.2998 9.10284 69.8768 7.59846 69.8494 6.04406C69.822 4.48965 69.1923 3.00657 68.093 1.90726C66.9937 0.807959 65.5106 0.178263 63.9562 0.150837C62.4018 0.123411 60.8974 0.700397 59.76 1.76024L35 26.5102L10.24 1.76024C9.10259 0.700397 7.59822 0.123411 6.04381 0.150837C4.4894 0.178263 3.00632 0.807959 1.90702 1.90726C0.807714 3.00657 0.178019 4.48965 0.150593 6.04406C0.123167 7.59846 0.700153 9.10284 1.75999 10.2402L30.76 39.2402Z" />
  </svg>
);

export const CollapseAction: FC<IconsProps> = (props) => (
  <ActionIcon icon={CollapseIcon} {...props} />
);

export const HandleAction: FC<IconsProps> = (props) => (
  <ActionIcon icon={HandleIcon} {...props} />
);

export const DeleteAction: FC<IconsProps> = (props) => (
  <ActionIcon icon={<DeleteFilled />} {...props} />
);

export const EditAction: FC<IconsProps> = (props) => (
  <ActionIcon icon={<EditFilled />} {...props} />
);
