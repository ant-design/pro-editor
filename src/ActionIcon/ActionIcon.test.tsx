import { SmileOutlined } from '@ant-design/icons';
import { render } from '@testing-library/react';
import ActionIcon from './index';

describe('ActionIcon', () => {
  it('可正常渲染', () => {
    const { container } = render(<ActionIcon icon={<SmileOutlined />} />);
    expect(container).toMatchSnapshot();
  });

  it('带有标题', async () => {
    const { container, findByTestId } = render(
      <ActionIcon icon={<SmileOutlined />} data-testid={'smile'} title={'标题'} />,
    );
    expect(container).toMatchSnapshot();
    const icon = await findByTestId('smile');
    expect(icon.className.includes('ant-btn-icon')).toBeTruthy();
    expect(icon.className.includes('ant-editor-icon')).toBeTruthy();
  });
});
