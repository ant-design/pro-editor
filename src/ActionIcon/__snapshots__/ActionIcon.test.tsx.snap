// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionIcon > 可正常渲染 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: color 600ms cubic-bezier(0.215, 0.61, 0.355, 1),scale 400ms cubic-bezier(0.215, 0.61, 0.355, 1),background-color 100ms cubic-bezier(0.215, 0.61, 0.355, 1);
  transition: color 600ms cubic-bezier(0.215, 0.61, 0.355, 1),scale 400ms cubic-bezier(0.215, 0.61, 0.355, 1),background-color 100ms cubic-bezier(0.215, 0.61, 0.355, 1);
}

.emotion-0:hover {
  color: rgba(0, 0, 0, 0.88)!important;
}

.emotion-0:active {
  scale: 0.8;
  color: rgba(0, 0, 0, 0.88);
}

<div>
  <button
    class="ant-btn ant-btn-text ant-btn-icon-only ant-editor-icon emotion-0"
    type="button"
  >
    <span
      class="ant-btn-icon"
    >
      <span
        aria-label="smile"
        class="anticon anticon-smile"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="smile"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M288 421a48 48 0 1096 0 48 48 0 10-96 0zm352 0a48 48 0 1096 0 48 48 0 10-96 0zM512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm263 711c-34.2 34.2-74 61-118.3 79.8C611 874.2 562.3 884 512 884c-50.3 0-99-9.8-144.8-29.2A370.4 370.4 0 01248.9 775c-34.2-34.2-61-74-79.8-118.3C149.8 611 140 562.3 140 512s9.8-99 29.2-144.8A370.4 370.4 0 01249 248.9c34.2-34.2 74-61 118.3-79.8C413 149.8 461.7 140 512 140c50.3 0 99 9.8 144.8 29.2A370.4 370.4 0 01775.1 249c34.2 34.2 61 74 79.8 118.3C874.2 413 884 461.7 884 512s-9.8 99-29.2 144.8A368.89 368.89 0 01775 775zM664 533h-48.1c-4.2 0-7.8 3.2-8.1 7.4C604 589.9 562.5 629 512 629s-92.1-39.1-95.8-88.6c-.3-4.2-3.9-7.4-8.1-7.4H360a8 8 0 00-8 8.4c4.4 84.3 74.5 151.6 160 151.6s155.6-67.3 160-151.6a8 8 0 00-8-8.4z"
          />
        </svg>
      </span>
    </span>
  </button>
</div>
`;

exports[`ActionIcon > 带有标题 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: color 600ms cubic-bezier(0.215, 0.61, 0.355, 1),scale 400ms cubic-bezier(0.215, 0.61, 0.355, 1),background-color 100ms cubic-bezier(0.215, 0.61, 0.355, 1);
  transition: color 600ms cubic-bezier(0.215, 0.61, 0.355, 1),scale 400ms cubic-bezier(0.215, 0.61, 0.355, 1),background-color 100ms cubic-bezier(0.215, 0.61, 0.355, 1);
}

.emotion-0:hover {
  color: rgba(0, 0, 0, 0.88)!important;
}

.emotion-0:active {
  scale: 0.8;
  color: rgba(0, 0, 0, 0.88);
}

<div>
  <button
    class="ant-btn ant-btn-text ant-btn-icon-only ant-editor-icon emotion-0"
    data-testid="smile"
    type="button"
  >
    <span
      class="ant-btn-icon"
    >
      <span
        aria-label="smile"
        class="anticon anticon-smile"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="smile"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M288 421a48 48 0 1096 0 48 48 0 10-96 0zm352 0a48 48 0 1096 0 48 48 0 10-96 0zM512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm263 711c-34.2 34.2-74 61-118.3 79.8C611 874.2 562.3 884 512 884c-50.3 0-99-9.8-144.8-29.2A370.4 370.4 0 01248.9 775c-34.2-34.2-61-74-79.8-118.3C149.8 611 140 562.3 140 512s9.8-99 29.2-144.8A370.4 370.4 0 01249 248.9c34.2-34.2 74-61 118.3-79.8C413 149.8 461.7 140 512 140c50.3 0 99 9.8 144.8 29.2A370.4 370.4 0 01775.1 249c34.2 34.2 61 74 79.8 118.3C874.2 413 884 461.7 884 512s-9.8 99-29.2 144.8A368.89 368.89 0 01775 775zM664 533h-48.1c-4.2 0-7.8 3.2-8.1 7.4C604 589.9 562.5 629 512 629s-92.1-39.1-95.8-88.6c-.3-4.2-3.9-7.4-8.1-7.4H360a8 8 0 00-8 8.4c4.4 84.3 74.5 151.6 160 151.6s155.6-67.3 160-151.6a8 8 0 00-8-8.4z"
          />
        </svg>
      </span>
    </span>
  </button>
</div>
`;
